### ISSUE TITLE
How to return html body on PATCH/POST/PUT?

### ISSUE THREAD

**Issue by javabr** [SoftInstigate/restheart#161]
I am trying to put together a solution with restheart and datatables - (HTML tables and forms  at https://editor.datatables.net/manual/server).

The only problem I could not solve elegantly on putting those 2 together is the fact that datables asks for a JSON body when PUT/POST/PATCH are returning from the server.

How I could do that? I looked at Hooks and Transformers in RestHeart but it does not seem like they can change the response body.

**Comment by ujibang** [comment]
Hi @javabr 

this is not currently possible because the write requests are not supposed to return data so their handlers chain do not include the ResponseSenderHandler (that is in charge of sending data back to the client).

However it indeed makes sense to allow a response transformer to define some response data; so I created the following improvement in our backlog https://softinstigate.atlassian.net/browse/RH-213 to be included in upcoming 3.0 release

I cannot commit on a release date but it should be by November.

**Comment by javabr** [comment]
Hi @ujibang , this is really great! Thank you for looking into this.

I think I can workaround my backlog and wait for this release. 

The datatable web page above has some examples of return data for each type of request - For example, it describes what would be the response json for POSTs when multiple docs vs single documents are pushed from the datatables to the server.

In my use cases I will never be sending more than one document - datatable has an option to disable multiple selections / multiple deletions/ multiple edits.

I will subscribe to the jira and put all my comments there. Although I understand this would not be a change specific for datatable clients, it is a good start and has real use cases to complement your analises and design.

**Comment by ujibang** [comment]
for you information I just pushed a change that address this need (response transformers can be applied to write requests). 

you can build it on your own or find the binaries from [sonatype](https://oss.sonatype.org/content/repositories/snapshots/org/restheart/restheart/3.0.0-SNAPSHOT/restheart-3.0.0-20161121.175236-21.tar.gz)

I also created a transformer that returns the written data together with the old data (in case of an update) in the response body; see the code at [WriteResultTransformer](https://github.com/SoftInstigate/restheart/blob/master/src/main/java/org/restheart/hal/metadata/singletons/WriteResultTransformer.java)

**Comment by javabr** [comment]
Big Thank you! it did work.

Just to document how to add this transformer to my transformer pipeline, I have created the collection with the following:
curl -i -X PUT \
   -H "Authorization:Basic *******" \
   -H "Content-Type:application/json" \
   -d \
'{"rts":[{"name":"responseTransformer", "phase":"RESPONSE", "scope":"CHILDREN"}]}' \
 'http://localhost:8080/osso/test3'

Then I have configured the transformer into the restheart.yml:
    - group: transformers
      interface: org.restheart.hal.metadata.singletons.Transformer
      singletons:
      ....... so many transformers here in between .....
        _**- name: responseTransformer
          class: org.restheart.hal.metadata.singletons.WriteResultTransformer**_


That is it! Now all my responses are decorated by WriteResultTransformer. I will probably write my on transformer latter on based on the WriteResultTransformer, but this one will do for now!
